src/memusage.h

```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
+// with multiple configurations (alignments, noexcept hash, node sizes).
+template <typename Key, typename V, typename Hash, typename Equals, typename Allocator>
+struct NodeSize<std::unordered_map<Key, V, Hash, Equals, Allocator>> {
+    [[nodiscard]] static constexpr size_t Value()
+    {
+        // libstdc++, libc++, and MSVC implement the nodes differently. To get the correct size
+        // with the correct alignment, we can simulate that with accordingly nested std::pairs.
+        using ValueType = std::pair<const Key, V>;
+
+#if defined(_MSC_VER)
+        // list node contains 2 pointers and no hash; see
+        // https://github.com/microsoft/STL/blob/main/stl/inc/unordered_map and
+        // https://github.com/microsoft/STL/blob/main/stl/inc/list
+        return sizeof(std::pair<std::pair<void*, void*>, ValueType>);
+#else
+
+#if defined(_LIBCPP_VERSION) // defined in any C++ header from libc++
+        // libc++ always stores hash and pointer in the node
+        // see https://github.com/llvm/llvm-project/blob/release/13.x/libcxx/include/__hash_table#L92
+        return sizeof(std::pair<ValueType, std::pair<size_t, void*>>);
+#else
+        // libstdc++ doesn't store hash when its operator() is noexcept;
+        // see hashtable_policy.h, struct _Hash_node
+        // https://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a05689.html
+        if (noexcept(std::declval<Hash>()(std::declval<const Key&>()))) {
+            return sizeof(std::pair<void*, ValueType>);
+        } else {
+            // hash is stored along ValueType, and that is wrapped with the pointer.
+            return sizeof(std::pair<void*, std::pair<ValueType, size_t>>);
+        }
+#endif
+#endif
+    }
 };
 
 template<typename X, typename Y>
 static inline size_t DynamicUsage(const std::unordered_set<X, Y>& s)
 {
-    return MallocUsage(sizeof(unordered_node<X>)) * s.size() + MallocUsage(sizeof(void*) * s.bucket_count());
+    return MallocUsage(sizeof(std::pair<X, void*>)) * s.size() + MallocUsage(sizeof(void*) * s.bucket_count());
 }
 
-template<typename X, typename Y, typename Z>
-static inline size_t DynamicUsage(const std::unordered_map<X, Y, Z>& m)
+template <typename Key, typename Value, typename Hash, typename Equals, typename Alloc>
+static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equals, Alloc>& m)
 {
-    return MallocUsage(sizeof(unordered_node<std::pair<const X, Y> >)) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
+    using ContainerType = std::unordered_map<Key, Value, Hash, Equals, Alloc>;
+    if constexpr (std::is_same_v<Alloc, node_allocator::Allocator<std::pair<const Key, Value>, memusage::NodeSize<ContainerType>::Value()>>) {
+        // Assumes that DynamicUsage of the MemoryResource is called separately. We don't do it here
+        // because multiple maps could use the same MemoryResource.
+        return MallocUsage(sizeof(void*) * m.bucket_count());
```

Comment:
I've played with this for a while now, but it doesn't work that way because in memusage.h I still need to know if the `node_allocator` is used for a container, because the node sizes are different. For the standard `std::unordered_map` I need to use `memusage::MallocUsage` for the nodes, and for `node_allocator` I have use the node size directly. So I need at least a forward declaration of `node_allocator::Allocator`. But that is fine I guess, I'll create a forward declaration then there's no circular dependency any more.

Also I'll split up `NodeSize` calculation into a separate independent file that both `memusage` and the allocator can use, so there are no circular dependency for that too.